source/MRViewer/MRSurfaceContoursWidget.cpp

```diff
@@ -558,13 +552,75 @@ void SurfaceContoursWidget::reset()
         FilterHistoryByCondition( [&] ( const std::shared_ptr<HistoryAction>& action )
         {
             bool res =
-                bool( std::dynamic_pointer_cast< AddPointActionPickerPoint >( action ) ) ||
-                bool( std::dynamic_pointer_cast< RemovePointActionPickerPoint >( action ) ) ||
-                bool( std::dynamic_pointer_cast< ChangePointActionPickerPoint >( action ) );
+                bool( std::dynamic_pointer_cast<AddPointActionPickerPoint>( action ) ) ||
+                bool( std::dynamic_pointer_cast<RemovePointActionPickerPoint>( action ) ) ||
+                bool( std::dynamic_pointer_cast<ChangePointActionPickerPoint>( action ) ) ||
+                bool( std::dynamic_pointer_cast<SurfaceContoursWidgetClearAction>( action ) );
             return res;
         } );
     }
 
     disconnect();
 }
-} // namespace MR 
\ No newline at end of file
+
+SurfaceContoursWidgetClearAction::SurfaceContoursWidgetClearAction( std::string name, SurfaceContoursWidget& widget )
+    : name_( std::move( name ) )
+    , widget_( widget )
+{
+    for ( const auto& [obj, contour] : widget_.pickedPoints_ )
+    {
+        ObjectState state;
+        state.objPtr = obj;
+        state.pickedPoints.reserve( contour.size() );
+        for ( const auto& p : contour )
+            state.pickedPoints.emplace_back( p->getCurrentPosition() );
+        states_.emplace_back( std::move( state ) );
+    }
+
+    activeObject_ = widget_.activeObject_;
+    activeIndex_ = widget_.activeIndex_;
+}
+
+void SurfaceContoursWidgetClearAction::action( Type type )
+{
+    if ( !widget_.isPickerActive_ )
+        return;
+
+    const auto prevWriteHistory = widget_.params.writeHistory;
+    widget_.params.writeHistory = false;
+    MR_FINALLY {
+        widget_.params.writeHistory = prevWriteHistory;
+    };
+
+    switch ( type )
+    {
+        case Type::Undo:
+            for ( const auto& state : states_ )
+            {
+                if ( const auto obj = state.objPtr.lock() )
+                {
+                    for ( const auto& p : state.pickedPoints )
+                        widget_.appendPoint( obj, p );
+                }
+            }
+            if ( const auto activeObject = activeObject_.lock() )
+            {
+                widget_.setActivePoint( activeObject, activeIndex_ );
+            }
+            break;
+
+        case Type::Redo:
+            widget_.clear();
+            break;
+    }
+}
+
+size_t SurfaceContoursWidgetClearAction::heapBytes() const
+{
+    return
+        name_.capacity()
+        + MR::heapBytes( states_ )
+    ;
+}
```

Comment:
actions that are supposed to be filtered usually return 0  here, not to be removed